src/detect-engine-alert.c

```diff
@@ -206,14 +200,18 @@ int PacketAlertAppend(DetectEngineThreadCtx *det_ctx, const Signature *s,
         p->alerts.alerts[p->alerts.cnt].frame_id =
                 (flags & PACKET_ALERT_FLAG_FRAME) ? det_ctx->frame_id : 0;
     } else {
-        /* We need to make room for this s->num
-         (a bit ugly with memcpy but we are planning changes here)*/
-        for (i = p->alerts.cnt - 1; i >= 0 && p->alerts.alerts[i].num > s->num; i--) {
-            memcpy(&p->alerts.alerts[i + 1], &p->alerts.alerts[i], sizeof(PacketAlert));
-        }
+        /* find position to insert */
+        for (i = p->alerts.cnt - 1; i >= 0 && p->alerts.alerts[i].num > s->num; i--)
+            ;
 
         i++; /* The right place to store the alert */
 
+        const uint16_t target_pos = i + 1;
+        const uint16_t space_post_target = packet_alert_max - 1 - i;
+        const uint16_t to_move = MIN(space_post_target, (p->alerts.cnt - i));
+        memmove(p->alerts.alerts + target_pos, p->alerts.alerts + i,
```

Comment:
Will this work correctly if `to_move == 0` ?

Reply:
I'll add a BUG_ON but given the check above it should not ever be 0

Reply:
> it should not ever be 0

It can be 0 : if `space_post_target` is 0 c-which can be the case if you have like `packet_alert_max`= 3, you have 3 alerts with num 3,5,7 and you want to add one with num 6

Than no memmove, just replace the last one


Otherwise, all is good :-)

Reply:
S-V tests failing with `BUG_ON(to_move == 0)` atm:
===> http2-bugfixes: FAILED: got exit code -6, expected 0
===> http2-files: FAILED: got exit code -6, expected 0
===> http-sticky-server: FAILED: got exit code -6, expected 0
===> smb2-07-frames: FAILED: got exit code -6, expected 0
===> tls13-draft28-frames: FAILED: got exit code -6, expected 0